### PR TITLE
Task 4. Add tests and e2e test

### PR DESCRIPTION
**Pull request by Gerasik**

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -3,7 +3,7 @@ import { BrowserRouter as Router, Switch, Route, Redirect } from 'react-router-d
 
 import Header from 'components/Header';
 import BodyContainer from 'containers/Body';
-import Search from 'components/Search';
+import SearchContainer from 'containers/Search';
```
Comment (shpakkdv) [inline src/App.tsx]: I think just `Search` is good. `SearchContainer` sounds a bit redundant.

File: src/components/ErrorBoundary/index.test.tsx [hunk 2]
```diff
@@ -0,0 +1,40 @@
+/* eslint-disable @typescript-eslint/camelcase */
+import React from 'react';
+import { shallow, mount } from 'enzyme';
+
+import ErrorBoundary from './index';
+
+class Buggy extends React.Component {
```
Comment (shpakkdv) [inline src/components/ErrorBoundary/index.test.tsx]: Why not just?
```tsx
const Buggy: React.FunctionComponent<{ throwError?: boolean }> = ({ throwError = true }) => {
  if (throwError) {
    throw new Error('An error has occurred in Buggy component!');
  }

  return null;
};
```

File: src/components/ErrorBoundary/index.test.tsx [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  constructor(props) {
+    super(props);
+    throw new Error('An error has occured in Buggy component!');
```
Comment (shpakkdv) [inline src/components/ErrorBoundary/index.test.tsx]: `occurred`

File: src/components/Footer/style.module.scss [hunk 4]
```diff
@@ -1,10 +1,10 @@
 footer:local {
-  background-color: $thirdColor;
+  background-color: $tertiaryColor;
```
Comment (shpakkdv) [inline src/components/Footer/style.module.scss:2]: `tertiary` is fancy name :)

File: src/containers/Body/selectors.ts [hunk 5]
```diff
@@ -4,8 +4,20 @@ import { createSelector } from 'reselect';
 import * as Models from 'models';
 import { State } from './models';
 
-const moviesResponse = (state): Models.MoviesResponseMap => (state as State).get('moviesResponse');
+const moviesResponse = (state): Models.MoviesResponseMap => {
+  return (state.bodyReducer as State).get('moviesResponse');
```
Comment (shpakkdv) [inline src/containers/Body/selectors.ts]: Why is not your `state` immutable? This is incorrect.

File: src/containers/Body/selectors.ts [hunk 6]
```diff
@@ -4,8 +4,20 @@ import { createSelector } from 'reselect';
...
-  return moviesResponse.get('data') || Immutable.List();
-});
+const searchSortBy = (state): string => {
+  return state.searchReducer.get('sortBy');
+};
+
+export const moviesData = createSelector(
+  moviesResponse,
+  searchSortBy,
+  (moviesResponse, searchSortBy) => ({
```
Comment (shpakkdv) [inline src/containers/Body/selectors.ts:18]: Better to operate immutable data everywhere. Such way you have inconsistency.

File: src/containers/Search/models.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  ) => Action.CHANGE_SEARCH_STRING;
+  export type CHANGE_SEARCH_BY = (payload: Payload.CHANGE_SEARCH_BY) => Action.CHANGE_SEARCH_BY;
+  export type CHANGE_SORT_BY = (payload: Payload.CHANGE_SORT_BY) => Action.CHANGE_SORT_BY;
+}
+
+export namespace Action {
+  export type CHANGE_SEARCH_STRING = Models.IAction<Payload.CHANGE_SEARCH_STRING>;
+  export type CHANGE_SEARCH_BY = Models.IAction<Payload.CHANGE_SEARCH_BY>;
+  export type CHANGE_SORT_BY = Models.IAction<Payload.CHANGE_SORT_BY>;
+}
```
Comment (shpakkdv) [inline src/containers/Search/models.ts:30]: Look at `containers/Body` and rewrite these types please in according to your own style
```ts
export namespace Payload {
  export type SetData = Models.MoviesResponse;
}

export namespace Action {
  export type SetData = Models.IAction<Payload.SetData>;
}

export namespace ActionCreator {
  export type SetData = (payload: Payload.SetData) => Action.SetData;
}
```

Also, `payload` is usually object.

**Review by shpakkdv — CHANGES_REQUESTED** [review]

File: src/containers/Body/actions.ts [hunk 8]
```diff
@@ -6,6 +6,10 @@ export const setData: ActionCreator.SetData = payload => ({
   payload,
 });
 
+export const setStartData: ActionCreator.SetStartData = () => ({
+  type: ActionType.SET_START_DATA,
+});
+
 // export const increment: Action. = payload => ({
```
Comment (shpakkdv) [inline src/containers/Body/actions.ts:13]: Why do you need it?

File: src/containers/Body/index.test.tsx [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+  });
+
+  it('should render with next state data', () => {
+    store.dispatch(setData(testMovieResponse));
+    expect(wrapper).toMatchSnapshot();
+  });
+});
+
+describe('Test reducer', () => {
+  // it('should return the initial state', () => {
```
Comment (shpakkdv) [inline src/containers/Body/index.test.tsx:53]: `it.skip`

File: src/containers/Body/models.ts [hunk 10]
```diff
@@ -12,8 +12,10 @@ export namespace Payload {
 
 export namespace Action {
   export type SetData = Models.IAction<Payload.SetData>;
+  export type SetStartData = { type: string };
```
Comment (shpakkdv) [inline src/containers/Body/models.ts:15]: `export type SetStartData = Models.IAction<void>;`

Or to do something like this:
```ts
export interface IEmptyAction {
  type: string;
}

export interface IAction<T> extends IEmptyAction {
  payload: T;
}
```

```ts
export type SetStartData = Models.IEmptyAction;
```

File: src/containers/Body/selectors.ts [hunk 11]
```diff
@@ -4,8 +4,20 @@ import { createSelector } from 'reselect';
 import * as Models from 'models';
 import { State } from './models';
 
-const moviesResponse = (state): Models.MoviesResponseMap => (state as State).get('moviesResponse');
+const moviesResponse = (state): Models.MoviesResponseMap => {
+  return (state.get('bodyReducer') as State).get('moviesResponse');
```
Comment (shpakkdv) [inline src/containers/Body/selectors.ts:8]: I think the word `Reducer` is redundant everywhere in names of store properties.

File: src/testData.ts [hunk 12]
```diff
@@ -0,0 +1,72 @@
+/* eslint-disable @typescript-eslint/camelcase */
+import { fromJS } from 'immutable';
+
+import { Movie, MoviesResponse, MoviesData, MovieMap } from 'models';
+import { State } from './containers/Body/models';
+
+export const testMovie1: Movie | MovieMap = {
```
Comment (shpakkdv) [inline src/testData.ts:7]: No `MovieMap`

File: src/testData.ts [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+  release_date: '2020-05-01',
+  poster_path: '/movie_url/image.jpg',
+  overview: 'ove',
+  budget: 54545,
+  revenue: 5454,
+  runtime: 5454,
+  genres: ['drama', 'sitcom', 'drama'],
+};
+
+export const testMovie2: Movie | MovieMap = {
```
Comment (shpakkdv) [inline src/testData.ts:22]: No `MovieMap`

File: src/testData.ts [hunk 14]
```diff
@@ -0,0 +1,72 @@
+/* eslint-disable @typescript-eslint/camelcase */
+import { fromJS } from 'immutable';
```
Comment (shpakkdv) [inline src/testData.ts:2]: These are mocks. You can rename into `data.mock.ts` or something similar with `.mock.ts` postfix

**Review by shpakkdv — APPROVED** [review]

File: src/containers/Body/models.ts (hunk 10 above)
Comment (Gerasik) [reply src/containers/Body/models.ts:15]: I have errors

**Pull request merged**